polkadot/node/core/approval-voting/src/criteria.rs

```diff
@@ -88,25 +91,105 @@ impl From<OurAssignment> for crate::approval_db::v1::OurAssignment {
 	}
 }
 
-fn relay_vrf_modulo_transcript(relay_vrf_story: RelayVRFStory, sample: u32) -> Transcript {
-	// combine the relay VRF story with a sample number.
-	let mut t = Transcript::new(approval_types::RELAY_VRF_MODULO_CONTEXT);
-	t.append_message(b"RC-VRF", &relay_vrf_story.0);
-	sample.using_encoded(|s| t.append_message(b"sample", s));
+// Combines the relay VRF story with a sample number if any.
+fn relay_vrf_modulo_transcript_inner(
+	mut transcript: Transcript,
+	relay_vrf_story: RelayVRFStory,
+	sample: Option<u32>,
+) -> Transcript {
+	transcript.append_message(b"RC-VRF", &relay_vrf_story.0);
 
-	t
+	if let Some(sample) = sample {
+		sample.using_encoded(|s| transcript.append_message(b"sample", s));
+	}
+
+	transcript
+}
+
+fn relay_vrf_modulo_transcript_v1(relay_vrf_story: RelayVRFStory, sample: u32) -> Transcript {
+	relay_vrf_modulo_transcript_inner(
+		Transcript::new(approval_types::v1::RELAY_VRF_MODULO_CONTEXT),
+		relay_vrf_story,
+		Some(sample),
+	)
+}
+
+fn relay_vrf_modulo_transcript_v2(relay_vrf_story: RelayVRFStory) -> Transcript {
+	relay_vrf_modulo_transcript_inner(
+		Transcript::new(approval_types::v2::RELAY_VRF_MODULO_CONTEXT),
+		relay_vrf_story,
+		None,
+	)
+}
+
+/// A hard upper bound on num_cores * target_checkers / num_validators
+const MAX_MODULO_SAMPLES: usize = 40;
+
+use std::convert::AsMut;
+
+fn clone_into_array<A, T>(slice: &[T]) -> A
+where
+	A: Default + AsMut<[T]>,
+	T: Clone,
+{
+	let mut a = A::default();
+	<A as AsMut<[T]>>::as_mut(&mut a).clone_from_slice(slice);
+	a
+}
+
+struct BigArray(pub [u8; MAX_MODULO_SAMPLES * 4]);
+
+impl Default for BigArray {
+	fn default() -> Self {
+		BigArray([0u8; MAX_MODULO_SAMPLES * 4])
+	}
+}
+
+impl AsMut<[u8]> for BigArray {
+	fn as_mut(&mut self) -> &mut [u8] {
+		self.0.as_mut()
+	}
+}
+
+/// Takes the VRF output as input and returns a Vec of cores the validator is assigned
+/// to as a tranche0 checker.
+fn relay_vrf_modulo_cores(
+	vrf_in_out: &VRFInOut,
+	// Configuration - `relay_vrf_modulo_samples`.
+	num_samples: u32,
+	// Configuration - `n_cores`.
+	max_cores: u32,
+) -> Vec<CoreIndex> {
+	if num_samples as usize > MAX_MODULO_SAMPLES {
+		gum::warn!(
+			target: LOG_TARGET,
+			n_cores = max_cores,
+			num_samples,
+			max_modulo_samples = MAX_MODULO_SAMPLES,
+			"`num_samples` is greater than `MAX_MODULO_SAMPLES`",
+		);
+	}
+
+	vrf_in_out
+		.make_bytes::<BigArray>(approval_types::v2::CORE_RANDOMNESS_CONTEXT)
+		.0
+		.chunks_exact(4)
+		.take(num_samples as usize)
+		.map(move |sample| CoreIndex(u32::from_le_bytes(clone_into_array(&sample)) % max_cores))
+		.unique()
+		.collect::<Vec<CoreIndex>>()
```

Comment:
This is sampling with deduplication, not with replacement.  It's better if we sample without replacement.
```
assert!(2 * num_samples < max_cores, "We enforce this in configuration at ... qed");
assert!(max_cores < u16::MAX as u32, "We enforce this in configuration at ... qed");

let mut rand_chacha::ChaChaRng = vrf_in_out.make_rng(approval_types::v2::CORE_RANDOMNESS_CONTEXT);
let mut res: Vec<CoreIndex> = Vec::with_capacity(max_cores);
while res.len() < num_samples {
    use rand::Rng;
    res.push(loop {
        let n: u32 = rng.gen_range(0..max_cores);
        if ! res.contains(n) { break n; }
    })
}
res
```

Reply:
`rand::seq::SliceRandom` has a couple methods too, but we're sampling from `0..max_cores` so this looks cleaner.

Reply:
A few clarification, before I go and update the implementation with your suggested change.

> This is sampling with replacement and deduplication. It's better if we sample without replacement.

I'm confused a bit here, the current implementation is sampling without `deduplication`  your suggested changes seem to do both, so what algorithm you propose we use ?


```
res.push(loop { let n: u32 = rng.gen_range(0..max_cores); if ! res.contains(n) { break n; } })
```

This will make our sampling algorithm have indeterministic computing time and potentially be slower on certain configurations. Bare in mind this is executed when we are checking the assignments as well, so all nodes would have to execute this for all assignments.

Can we actually use  a max number of iterations for this loop  ?

On using `ChaChaRng`, I'm not an expert here, but isn't  there a risk here where some configurations change behind the scene and might end up in a situation where we can't replicate the same random numbers on the node checking the assignments.


>With the ChaCha algorithm it is possible to choose the number of rounds the core algorithm should run. The number of rounds is a tradeoff between performance and security, where 8 rounds is the minimum potentially secure configuration, and 20 rounds is widely used as a conservative choice. We use 20 rounds in this implementation, but hope to allow type-level configuration in the future

Just flipping through the documentation it seems there are some knobs that might change in the future, without us actually observing.

So, imagine a world where they decide to change the number of rounds, then when we deploy the network it would be split between old and new behaviour, so this might end up slowing/stalling the network because nodes won't be able to check each other assignments. Probability of this happening is probably really really low, but I still don't think it is worth taking the risk in perpetuity, wouldn't you agree ?

Reply:
We should "sample without replacement" which means we produce exactly `num_samples` samples, all of which are distinct.

At present you "sample with replacement" which yields duplicates, but then remove the duplicates, so your code produces fewer than `num_samples` in general.  It's only a small difference, assuming many cores, but it's one less thing that's slightly different between the code and the theory.

Reply:
Yes, fine use https://rust-random.github.io/rand/rand_chacha/struct.ChaCha20Rng.html instead.

If you enforce my assert somewhere else, then the loop terminates with probability >1/2 in each iteration (or > 90% or > 95% with production parameters). 

`rand::seq::SliceRandom::partial_shuffle` can do this with exactly `num_samples` iterations, but only after you first build a workspace array `(0..num_cores).collect::<Vec<CoreIndex>>();`.

At these sizes the cache misses from building the array should out weight the extra chacha outputs, but whatever.

Reply:
Addressed this with https://github.com/paritytech/polkadot-sdk/pull/1178/commits/c8eff61fc50bf900e39ab44727f1b45bbf960b0f.
 
 @burdges let me know if there is anything else  that you think needs addressing on this PR, otherwise I'll go ahead with merging it.